canonical/maas-ui
# feat(machines): handle bulk actions results

## Done

- handle bulk machine actions results (failed system id count and success count)
- add `useSelectedMachinesActionsDispatch` hook for handling of multiple action dispatches when there are groups and items selected
- stricter type checking for machine slice reducer with `MachineActionStatus`
- refactor to use `selectedMachines` instead of selectedFilter (initially I wanted to simplify this, but because we need to handle multiple dispatches in case there are selected both groups and items this was not possible)

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to machine listing
- Select a few machines and dispatch an action via "Take action" menu
- Ensure that the form is closed upon successful submission
- If you select a group of machines and an item outside of that group, there should be 2 machine.action websocket calls - 1 for the groups, 1 for the items
- If an action is not successful for all selected nodes, the form should remain open and an error message should be displayed
- if the back-end responds with a failed system ids an error message indicating the number of failed nodes should be displayed
  - you can test this by testing a machine and attempting to test it again once testing is in progress

## Fixes

Fixes: https://github.com/canonical/app-tribe/issues/1397
Fixes: https://github.com/canonical/app-tribe/issues/1396

## Launchpad issue

Related Launchpad maas issue in the form `lp#number`.

## Backports

In general, please propose fixes against _main_ rather than release branches (e.g. 2.7), unless the fix is only applicable for that specific release. Please apply backport labels to the PR (e.g. "Backport 2.7") for the appropriate releases to target.

Only bug and security fixes should be backported, new features should only land in main.

## Screenshots

It could be helpful to provide some screenshots to aid in QAing the change.
